packages/core/src/actions/execute.ts

```diff
@@ -750,28 +1006,241 @@ export const runEntireDeploymentProcess = async (
+ * writing information about a deployment to the website backend is implemented in the `deploymentContext`.
+ *
+ * See documentation on the `DeploymentContext` type at the top of this file for more information of the specific fields
+ * and justification for why the logic implemented in them cannot be shared.
+ */
+export const compileAndExecuteDeployment = async (
+  deploymentContext: DeploymentContext
+): Promise<
+  | {
+      receipts: SphinxTransactionReceipt[]
+      batches: SphinxLeafWithProof[][]
+      finalStatus: BigInt
+      failureAction: HumanReadableAction | undefined
+    }
+  | undefined
+> => {
+  const { deployment } = deploymentContext
+  const networkName = deployment.networkName
+  const deploymentId = deployment.id
+  const rawCompilerConfigs = deployment.compilerConfigs
+  const { logger } = deploymentContext
+
+  if (!rawCompilerConfigs) {
+    deploymentContext.throwError(
+      `[Executor ${networkName}]: No compiler config stored for this deployment, should never happen`,
+      deployment.id,
+      networkName
+    )
+  }
+
+  logger?.info(`[Executor ${networkName}]: retrieving the deployment...`)
+  // Compile the bundle using either the provided localDeploymentId (when running the in-process
+  // executor), or using the Config URI
+  let merkleTree: SphinxMerkleTree
+  let compilerConfigs: Array<CompilerConfig>
+  let configArtifacts: ConfigArtifacts
+
+  // Handle if the config cannot be fetched
+  try {
+    ;({ merkleTree, compilerConfigs, configArtifacts } =
+      await buildDeploymentWithCompilerConfigs(rawCompilerConfigs!))
+  } catch (e: any) {
+    await deploymentContext.handleError(e, deployment, networkName, [])
+
+    return
+  }
+
+  const targetNetworkConfig = compilerConfigs.find(
+    (config) => config.chainId === deployment.chainId.toString()
+  )
+  if (!targetNetworkConfig) {
+    await deploymentContext.throwError(
+      `[Executor ${networkName}]: Error could not find target network config. This should never happen please report it to the developers.`,
+      deploymentId,
+      networkName
+    )
+    return
+  }
+
+  const { projectName } = targetNetworkConfig.newConfig
+
+  // get active deployment ID for this project
+  const moduleAddress = deployment.moduleAddress
+  const module = new ethers.Contract(
+    moduleAddress,
+    SphinxModuleABI,
+    deploymentContext.provider
+  )
+
+  if ((await deploymentContext.provider.getCode(moduleAddress)) !== '0x') {
+    const activeRoot = await module.activeMerkleRoot()
+    const deploymentStatus = await module.merkleRootStates(merkleTree.root)
+
+    if (
+      activeRoot === ethers.ZeroHash &&
+      deploymentStatus.numLeaves !== BigInt(0) &&
+      deploymentStatus.numLeaves === deploymentStatus.leavesExecuted
+    ) {
+      await deploymentContext.handleAlreadyExecutedDeployment(
+        deploymentContext,
+        targetNetworkConfig,
+        configArtifacts,
+        networkName
+      )
+
+      return
+    }
+  }
```

Comment:
Yeah, we want the website to be able to recover from any possible failure, including one where we failed to update the status of the deployment in the DB properly. So, we use the real on-chain state as a source of truth to determine whether the deployment was completed. 

For the same reason, when we later handle if the deployment failed. We do that by checking the on-chain state instead of relying fully on the data in the DB.